src/collision_viewer/main.cpp

```diff
@@ -0,0 +1,590 @@
+
+    std3D_pD3Device->SetRenderState(D3DRENDERSTATE_ZENABLE, 1);
+
+    viewport->Release();
+}
+
+void swrModel_UnkDraw_Hook(int x)
+{
+    if (!global_state || !global_state->enabled)
+    {
+        swrModel_UnkDraw(x);
+        return;
+    }
+
+    auto* root_node = swrModel_unk_array[x].model_root_node;
+    std::vector<swrModel_Node*> temp_children(root_node->child_nodes, root_node->child_nodes + root_node->num_children);
+
+    // first render the terrain...
+    if (global_state->settings.show_visual_mesh)
+    {
+        for (int i = 4; i < root_node->num_children; i++)
+            root_node->child_nodes[i] = NULL;
+
+        swrModel_UnkDraw(x);
+    }
+    else
+    {
+        stdDisplay_BackBufferFill(0, 0, 0, nullptr);
+    }
+
+    rdCache_Flush();
+    render_collision_meshes();
+
+    // ... and after the collision meshes render the rest:
+    for (int i = 0; i < root_node->num_children; i++)
+        root_node->child_nodes[i] = i < 4 ? NULL : temp_children[i];
+
+    swrModel_UnkDraw(x);
+
+    // restore node:
+    std::copy(temp_children.begin(), temp_children.end(), root_node->child_nodes);
+}
+
+void detour_attach(void** pPointer, void* pDetour, int num_bytes_to_copy)
+{
+    if (num_bytes_to_copy < 5)
+        abort();
+
+    uint8_t* original_address = (uint8_t*)*pPointer;
+    uint8_t* new_address = (uint8_t*)pDetour;
+    DWORD old_protect;
+    VirtualProtect(original_address, num_bytes_to_copy, PAGE_EXECUTE_READWRITE, &old_protect);
+
+    int32_t offset = new_address - (original_address + 5);
+
+    uint8_t* patch_memory = (uint8_t*)VirtualAlloc(nullptr, num_bytes_to_copy + 5, MEM_COMMIT, PAGE_EXECUTE_READWRITE);
+    memcpy(patch_memory, original_address, num_bytes_to_copy);
+
+    original_address[0] = 0xe9;
+    memcpy(original_address + 1, &offset, 4);
+
+    int32_t patch_offset = (original_address + num_bytes_to_copy) - (patch_memory + num_bytes_to_copy + 5);
+
+    patch_memory[num_bytes_to_copy] = 0xe9;
+    memcpy(patch_memory + num_bytes_to_copy + 1, &patch_offset, 4);
+
+    *pPointer = patch_memory;
+}
+
+void detour_detach(void** pPointer, void* pDetour, int num_bytes_to_copy)
+{
+    if (num_bytes_to_copy < 5)
+        abort();
+
+    uint8_t* patch_memory = (uint8_t*)*pPointer;
+
+    uint32_t patch_offset;
+    memcpy(&patch_offset, patch_memory + num_bytes_to_copy + 1, 4);
+    uint8_t* original_address = patch_offset + (patch_memory + num_bytes_to_copy + 5) - num_bytes_to_copy;
+
+    memcpy(original_address, patch_memory, num_bytes_to_copy);
+    VirtualFree(patch_memory, num_bytes_to_copy + 5, MEM_RELEASE);
+
+    DWORD old_protect;
+    VirtualProtect(original_address, num_bytes_to_copy, PAGE_EXECUTE_READ, &old_protect);
+
+    *pPointer = original_address;
+}
+
+extern "C" void init_collision_viewer(CollisionViewerState* global_state)
+{
+    ::global_state = global_state;
+    detour_attach((void**)&swrModel_UnkDraw, (void*)swrModel_UnkDraw_Hook, 5);
```

Comment:
the only thing that you have to be careful about is the "num_bytes_to_copy" parameter. you have to look at the disassembly of the specific function to know how many bytes need to be copied, to avoid slicing an instruction in half